homeassistant/components/sun.py

```diff
@@ -51,172 +51,143 @@ def is_on(hass, entity_id=None):
     return hass.states.is_state(entity_id, STATE_ABOVE_HORIZON)
 
 
-def next_dawn(hass, entity_id=None):
+def next_dawn(hass, entity_id=None, offset=None):
     """Local datetime object of the next dawn.
 
     Async friendly.
     """
-    utc_next = next_dawn_utc(hass, entity_id)
+    utc_next = next_dawn_utc(hass, entity_id, offset)
 
     return dt_util.as_local(utc_next) if utc_next else None
 
 
-def next_dawn_utc(hass, entity_id=None):
+def next_dawn_utc(hass, entity_id=None, offset=None):
     """UTC datetime object of the next dawn.
 
     Async friendly.
     """
-    entity_id = entity_id or ENTITY_ID
-
-    state = hass.states.get(ENTITY_ID)
-
-    try:
-        return dt_util.parse_datetime(
-            state.attributes[STATE_ATTR_NEXT_DAWN])
-    except (AttributeError, KeyError):
-        # AttributeError if state is None
-        # KeyError if STATE_ATTR_NEXT_DAWN does not exist
+    entity = hass.data.get(entity_id or ENTITY_ID)
+    if entity is None:
         return None
 
+    return Sun.get_next_solar_event(
```

Comment:
So the problem we're facing that's causing duplicate triggers, is that the next sunrise and next sunset information stored on the state machine is not sufficient information to determine the next event once offsets are added to the mix, since the correct time might need to be calculated as an offset relative to a sunset/sunrise on a different day. These are two alternate proposals that come to mind:

1) We have the sun automation platform directly invoke astral, creating it's own location object [similar to the sun component logic](https://github.com/armills/home-assistant/blob/f90c2de321aa21c8b8ce54d45e0a367c7a76839e/homeassistant/components/sun.py#L224). This would decouple the sun automation platform from the sun component. (This is my preference.)

2) Instead of only storing the next sunrise, and next sunset in the state machine, we store sunrise/sunset events for yesterday, today, and tomorrow. This should be sufficient information for both the sun automation condition and triggers to calculate the correct times that they need.